servers/src/main/java/tachyon/worker/block/BlockMetadataManager.java

```diff
@@ -0,0 +1,336 @@
+    int level = getTier(tierAlias).getTierLevel();
+    return mTiers.subList(level + 1, mTiers.size());
+  }
+
+  /**
+   * Gets the amount of available space in given location in bytes.
+   *
+   * @param location location the check available bytes
+   * @return available bytes
+   */
+  public synchronized long getAvailableBytes(BlockStoreLocation location) throws IOException {
+    long spaceAvailable = 0;
+
+    if (location.equals(BlockStoreLocation.anyTier())) {
+      for (StorageTier tier : mTiers) {
+        spaceAvailable += tier.getAvailableBytes();
+      }
+      return spaceAvailable;
+    }
+
+    int tierAlias = location.tierAlias();
+    StorageTier tier = getTier(tierAlias);
+    // TODO: This should probably be max of the capacity bytes in the dirs?
+    if (location.equals(BlockStoreLocation.anyDirInTier(tierAlias))) {
+      return tier.getAvailableBytes();
+    }
+
+    int dirIndex = location.dir();
+    StorageDir dir = tier.getDir(dirIndex);
+    return dir.getAvailableBytes();
+  }
+
+  /**
+   * Checks if the storage has a given block.
+   *
+   * @param blockId the block ID
+   * @return true if the block is contained, false otherwise
+   */
+  public synchronized boolean hasBlockMeta(long blockId) {
+    for (StorageTier tier : mTiers) {
+      for (StorageDir dir : tier.getStorageDirs()) {
+        if (dir.hasBlockMeta(blockId)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Gets the metadata of a block given its blockId or throws IOException.
+   *
+   * @param blockId the block ID
+   * @return metadata of the block or null
+   * @throws IOException if no BlockMeta for this blockId is found
+   */
+  public synchronized BlockMeta getBlockMeta(long blockId) throws IOException {
+    for (StorageTier tier : mTiers) {
+      for (StorageDir dir : tier.getStorageDirs()) {
+        if (dir.hasBlockMeta(blockId)) {
+          return dir.getBlockMeta(blockId);
+        }
+      }
+    }
+    throw new IOException("Failed to get BlockMeta: blockId " + blockId + " not found");
+  }
+
+  /**
+   * Moves the metadata of an existing block to another location or throws IOExceptions.
+   *
+   * @param blockMeta the meta data of the block to move
+   * @param newLocation new location of the block
+   * @return the new block metadata if success, absent otherwise
+   * @throws IOException if this block is not found
+   */
+  public synchronized BlockMeta moveBlockMeta(BlockMeta blockMeta, BlockStoreLocation newLocation)
+      throws IOException {
+    // If move target can be any tier, then simply return the current block meta.
+    // TODO: change this to belongTo
+    if (newLocation.equals(BlockStoreLocation.anyTier())) {
+      return blockMeta;
+    }
+
+    int newTierAlias = newLocation.tierAlias();
+    StorageTier newTier = getTier(newTierAlias);
+    StorageDir newDir = null;
+    if (newLocation.equals(BlockStoreLocation.anyDirInTier(newTierAlias))) {
```

Comment:
Maybe check if the `newTierAlias` is the same as the current tier the block is in?